### DISCUSSION TITLE
Differences Between Packaging IO Connectors and Functions in .jar vs .nar

### DISCUSSION THREAD

**Discussion by NurramoX** [apache/pulsar#22711] [Q&A] [answered]
Hi everyone,

I'm seeking some clarity on the differences between packaging IO connectors and functions in .jar or .nar formats within Pulsar.

Specifically, I'm interested in 
- Use Cases: When should one prefer .jar over .nar and vice versa for connectors and functions?
- Benefits and Limitations: What are the benefits or limitations associated with each packaging format?
- Best Practices: Are there any best practices or guidelines for choosing between these formats for different scenarios?

Thanks in advance for your insights!

**Comment by lhotari** [comment, marked as answer]
A nar file contains the dependent libraries in the file in META-INF/bundled-dependencies directory. Pulsar extracts the Nar file to a temporary directory using [NarUnpacker](https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarUnpacker.java) and uses [NarClassLoader](https://github.com/apache/pulsar/blob/master/pulsar-common/src/main/java/org/apache/pulsar/common/nar/NarClassLoader.java) to add all of the jar files in META-INF/bundled-dependencies to the classloader used to load a IO connector or function.

The alternative to a nar file would be to use a fat jar where all dependencies are combined in a single jar file.

The nar file doesn't provide classloader isolation from the parent classloader used in Pulsar. It's not possible to override a class that is available in the parent classloader when using Nar files. This could impact the behavior in some cases since there are multiple ways in Pulsar how Pulsar IO connectors or functions can be run.

One possible way to workaround that is to use a shaded jar instead of a nar file where all dependencies are relocated to a different package.

**Reply by NurramoX** [reply]
I see, that clarified the differences for me. Thank you for your response! And thank you for responding so quickly!
